src/t8_geometry/t8_geometry_implementations/t8_geometry_mapping.hxx

```diff
@@ -0,0 +1,113 @@
+   * is viable with all tree types and uses as many vertices as the tree type has.
+   * The vertices are saved via the \ref t8_cmesh_set_tree_vertices function.
+   * \param [in] dimension  0 <= \a dimension <= 3. Element dimension in the parametric space.
+   * \param [in] name       Element name.
+   */
+  finite_element (int dimension, const char *name);
+
+  /**
+   * Maps points in the reference space \f$ [0,1]^\mathrm{dim} \to \mathbb{R}^3 \f$.
+   * \param [in]  cmesh       The cmesh in which the point lies.
+   * \param [in]  gtreeid     The global tree (of the cmesh) in which the reference point is.
+   * \param [in]  ref_coords  Array of \a dimension x \a num_coords many entries, specifying points in \f$ [0,1]^\mathrm{dim} \f$.
+   * \param [in]  num_coords  Amount of points of /f$ \mathrm{dim} /f$ to map.
+   * \param [out] out_coords  The mapped coordinates in physical space of \a ref_coords. The length is \a num_coords * 3.
+   */
+  void
+  t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreeid, const double *ref_coords, const size_t num_coords,
+                    double out_coords[3]) const;
+
+  /**
+   * Compute the jacobian of the \a t8_geom_evaluate map at a point in the reference space \f$ [0,1]^\mathrm{dim} \f$.
+   * \param [in]  cmesh      The cmesh in which the point lies.
+   * \param [in]  gtreeid    The global tree (of the cmesh) in which the reference point is.
+   * \param [in]  ref_coords  Array of \a dimension x \a num_coords many entries, specifying points in \f$ [0,1]^\mathrm{dim} \f$.
+   * \param [in]  num_coords  Amount of points of /f$ \mathrm{dim} /f$ to map.
+   * \param [out] jacobian    The jacobian at \a ref_coords. Array of size \a num_coords x dimension x 3. Indices \f$ 3 \cdot i\f$ , \f$ 3 \cdot i+1 \f$ , \f$ 3 \cdot i+2 \f$
+   *                          correspond to the \f$ i \f$-th column of the jacobian  (Entry \f$ 3 \cdot i + j \f$ is \f$ \frac{\partial f_j}{\partial x_i} \f$).
+   */
+  virtual void
+  t8_geom_evaluate_jacobian (t8_cmesh_t cmesh, t8_gloidx_t gtreeid, const double *ref_coords, const size_t num_coords,
+                             double *jacobian) const;
+
+  virtual const std::vector<double>
+  basis (const double *ref_coords) const = 0;
+
+  /* Load tree data is inherited from t8_geometry_with_vertices. */
+};
+
+/* Three-node linear triangle */
+class T3: public finite_element {
+ public:
+  T3 (): finite_element (2, "Triangle-3")
+  {
+  }
+
+  const std::vector<double>
+  basis (const double *ref_coords) const override;
+};
+
+/* Six-node quadratic triangle */
+class T6: public finite_element {
+ public:
+  T6 (): finite_element (2, "Triangle-6")
+  {
+  }
+
+  const std::vector<double>
+  basis (const double *ref_coords) const override;
+};
+
+/* Four-node bilinear quadrilateral */
+class Q4: public finite_element {
+ public:
+  Q4 (): finite_element (2, "Quadrilateral-4")
+  {
+  }
+
+  const std::vector<double>
+  basis (const double *ref_coords) const override;
+};
```

Comment:
I really like the structural thinking behind this approach, but it collides with the principle of how our geometries work: Generally, each tree can get its own geometry, but this is not the expected way to use them. Normally, every tree gets the same geometry and then the geometry itself looks at the tree shape and decides how to map in that tree shape (just like the [linear geometry](https://github.com/DLR-AMR/t8code/blob/079956ff6115a0e956192d90d56c5f5cfa6e321f/src/t8_geometry/t8_geometry_helpers.c#L82) does it). So your geometry should have one `t8_geom_evaluate` and since this function gets the `active_tree_class` (the [shape enum](https://github.com/DLR-AMR/t8code/blob/079956ff6115a0e956192d90d56c5f5cfa6e321f/src/t8_eclass.h#L39)) of the currently loaded tree, you can use different basis functions for the different tree shapes. Since the basis functions can use a different amount of vertices according to their polynomial degree, I suggest saving a polynomial degree with each tree.
For that, you can use the function `t8_cmesh_set_attribute` to set an attribute for each tree during the definition of the cmesh and the function `t8_cmesh_get_attribute` to retrieve it during the geometry computation. [Here](https://github.com/DLR-AMR/t8code/blob/079956ff6115a0e956192d90d56c5f5cfa6e321f/src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.cxx#L140-L148) is an example on how to reimplement the `t8_geom_load_tree_data` function for that. You can define a new key for that in the [`t8_cmesh_types.h`](https://github.com/DLR-AMR/t8code/blob/main/src/t8_cmesh/t8_cmesh_types.h) header.

Reply:
> Since the basis functions can use a different amount of vertices according to their polynomial degree, I suggest saving a polynomial degree with each tree.

When I chose the current approach, I really thought of this mapping as a general finite element mapping. So I had in the back of my mind the possibility of using less conventional mappings (serendipity elements, elements with non-equal order interpolation, etc.). Thinking more about it, I have never met such bases for the *geometrical mapping*, so the polynomial degree will uniquely determine the basis functions for a given tree type.

Reply:
For these more unconventional mappings, one could use an enum defining these as negative polynomial degree integers. Or one could implement a more specialized, derived geometry for that